File: interflow-backend/src/models/reception.py
from pydantic import BaseModel, Field, field_validator
from models.matieres import Matiere
from datetime import datetime
from enum import Enum
from typing import List, Dict, Any, Optional, Iterator


class EtatReception(Enum):
    EN_COURS = "en_cours"
    TERMINEE = "terminée"
    ANNULEE = "annulée"
    RELACHE = "relâché"
    EN_ATTENTE = "en_attente"

class TypeReception(Enum):
    PRESTATAIRE = "prestataire"
    INTERNE = "interne"

class Reception(BaseModel):
    id: Optional[str] = Field(None, description="Identifiant unique de la réception (généré automatiquement)")
    type: TypeReception = Field(TypeReception.PRESTATAIRE, description="Type de réception")
    matiere: Matiere = Field(..., description="Matière réceptionnée")
    quantite: float = Field(..., description="Quantité réceptionnée", ge=0)
    lot: Optional[str] = Field("", description="Numéro de lot")
    qualification: bool = Field(False, description="Qualification requise")
    date_creation: datetime = Field(..., description="Date de création de la réception")
    date_modification: Optional[datetime] = Field(None, description="Date de dernière modification")

    # Champs spécifiques aux réceptions internes
    ordre: Optional[str] = Field(None, description="Numéro d'ordre")
    type_ordre: Optional[str] = Field(None, description="Type d'ordre")
    statut_ordre: Optional[str] = Field(None, description="Statut de l'ordre")
    poste: Optional[str] = Field(None, description="Poste de travail")
    statut_poste: Optional[str] = Field(None, description="Statut du poste")
    division: Optional[str] = Field(None, description="Division")
    magasin: Optional[str] = Field(None, description="Magasin")
    article: Optional[str] = Field(None, description="Code article")
    libelle_article: Optional[str] = Field(None, description="Libellé de l'article")
    ref_externe: Optional[str] = Field(None, description="Référence externe")
    description_externe: Optional[str] = Field(None, description="Description externe")
    quantite_ordre: Optional[float] = Field(None, description="Quantité de l'ordre")
    quantite_receptionnee: Optional[float] = Field(None, description="Quantité réceptionnée")
    udm: Optional[str] = Field(None, description="Unité de mesure")
    fournisseur: Optional[str] = Field(None, description="Fournisseur")
    description_fournisseur: Optional[str] = Field(None, description="Description du fournisseur")
    date_reception: Optional[datetime] = Field(None, description="Date de réception")

    # État unifié (pour compatibilité)
    etat: EtatReception = Field(EtatReception.EN_COURS, description="État de la réception")

    @field_validator('quantite_ordre', 'quantite_receptionnee', 'quantite', mode='before')
    @classmethod
    def normalize_numeric_values(cls, v):
        """Normalise les valeurs numériques (convertit les virgules en points)"""
        if v is None:
            return None if 'quantite_ordre' in cls.__name__ or 'quantite_receptionnee' in cls.__name__ else 0.0

        # Si c'est déjà un nombre, le garder
        if isinstance(v, (int, float)):
            return float(v)

        value_str = str(v).strip()

        # Convertir les virgules en points
        if ',' in value_str:
            # Gérer les cas spéciaux comme "500," -> "500.0"
            if value_str.endswith(','):
                value_str = value_str[:-1] + '.0'
            else:
                value_str = value_str.replace(',', '.')

        try:
            return float(value_str)
        except ValueError:
            return None if 'quantite_ordre' in cls.__name__ or 'quantite_receptionnee' in cls.__name__ else 0.0

    @field_validator('ordre', 'type_ordre', 'statut_ordre', 'poste', 'statut_poste', 'division',
                    'magasin', 'article', 'libelle_article', 'ref_externe', 'description_externe',
                    'udm', 'fournisseur', 'description_fournisseur', 'lot', mode='before')
    @classmethod
    def clean_string_fields(cls, v):
        """Nettoie les champs string pour gérer les valeurs NaN"""
        if v is None:
            return None

        # Si c'est déjà une string valide, la garder
        if isinstance(v, str):
            value_str = v.strip()
            if value_str.lower() in ['nan', 'none', 'null', '']:
                return None
            return value_str

        # Si c'est un float NaN, retourner None
        if isinstance(v, float) and str(v).lower() == 'nan':
            return None

        # Convertir en string et nettoyer
        value_str = str(v).strip()
        if value_str.lower() in ['nan', 'none', 'null', '']:
            return None

        return value_str

    def __init__(self, **kwargs):
        super().__init__(**kwargs)

        # Générer l'id si non fourni
        if not self.id:
            if self.type == TypeReception.INTERNE and self.ordre and self.article:
                # Utiliser ordre + article + date_reception + poste (sans heure)
                if self.date_reception:
                    date_str = self.date_reception.strftime("%Y%m%d")
                    poste_str = self.poste if self.poste else "NOPOSTE"
                    self.id = f"{self.ordre}_{self.article}_{date_str}_{poste_str}"
                else:
                    # Si pas de date de réception, utiliser la date de création
                    date_str = self.date_creation.strftime("%Y%m%d")
                    poste_str = self.poste if self.poste else "NOPOSTE"
                    self.id = f"{self.ordre}_{self.article}_{date_str}_{poste_str}"
            else:
                # Pour les réceptions prestataires, utiliser un id auto-généré
                import uuid
                self.id = str(uuid.uuid4())

        # Convertir les états de réceptions internes vers l'état unifié
        if self.type == TypeReception.INTERNE and self.statut_ordre:
            if self.statut_ordre.upper() in ["RELÂCHÉ", "RELACHE"]:
                self.etat = EtatReception.RELACHE
            elif self.statut_ordre.upper() in ["EN_ATTENTE", "EN ATTENTE"]:
                self.etat = EtatReception.EN_ATTENTE
            elif self.statut_ordre.upper() in ["TERMINÉ", "TERMINE"]:
                self.etat = EtatReception.TERMINEE
            elif self.statut_ordre.upper() in ["ANNULÉ", "ANNULE"]:
                self.etat = EtatReception.ANNULEE
            else:
                self.etat = EtatReception.EN_COURS

    def model_dump(self) -> Dict[str, Any]:
        """Sérialise le modèle en dictionnaire"""
        return {
            "id": self.id,
            "type": self.type.value if self.type else None,
            "matiere": self.matiere.model_dump() if self.matiere else None,
            "quantite": self.quantite,
            "lot": self.lot,
            "qualification": self.qualification,
            "date_creation": self.date_creation.isoformat() if self.date_creation else None,
            "date_modification": self.date_modification.isoformat() if self.date_modification else None,
            "etat": self.etat.value if self.etat else None,
            # Champs spécifiques aux réceptions internes
            "ordre": self.ordre,
            "type_ordre": self.type_ordre,
            "statut_ordre": self.statut_ordre,
            "poste": self.poste,
            "statut_poste": self.statut_poste,
            "division": self.division,
            "magasin": self.magasin,
            "article": self.article,
            "libelle_article": self.libelle_article,
            "ref_externe": self.ref_externe,
            "description_externe": self.description_externe,
            "quantite_ordre": self.quantite_ordre,
            "quantite_receptionnee": self.quantite_receptionnee,
            "udm": self.udm,
            "fournisseur": self.fournisseur,
            "description_fournisseur": self.description_fournisseur,
            "date_reception": self.date_reception.isoformat() if self.date_reception else None
        }

    @classmethod
    def from_model_dump(cls, data: Dict[str, Any]) -> 'Reception':
        """
        Désérialise un dictionnaire en modèle, avec gestion des valeurs manquantes

        Args:
            data: Le dictionnaire à désérialiser

        Returns:
            Reception: L'instance de Reception créée
        """
        # Copie du dictionnaire pour éviter de modifier l'original
        processed_data = data.copy()

        # Valeurs par défaut pour les champs manquants
        if "id" not in processed_data:
            processed_data["id"] = None

        if "type" not in processed_data:
            processed_data["type"] = TypeReception.PRESTATAIRE
        elif isinstance(processed_data["type"], str):
            try:
                processed_data["type"] = TypeReception(processed_data["type"])
            except ValueError:
                processed_data["type"] = TypeReception.PRESTATAIRE

        if "quantite" not in processed_data:
            processed_data["quantite"] = 0.0

        if "lot" not in processed_data:
            processed_data["lot"] = ""

        if "qualification" not in processed_data:
            processed_data["qualification"] = False

        if "date_creation" not in processed_data:
            processed_data["date_creation"] = datetime.now().replace(tzinfo=None)
        elif isinstance(processed_data["date_creation"], str):
            try:
                date_creation = datetime.fromisoformat(processed_data["date_creation"])
                # S'assurer que la date est naive (sans timezone) pour éviter les erreurs de comparaison
                if date_creation.tzinfo is not None:
                    date_creation = date_creation.replace(tzinfo=None)
                processed_data["date_creation"] = date_creation
            except ValueError:
                processed_data["date_creation"] = datetime.now().replace(tzinfo=None)

        if "date_modification" not in processed_data:
            processed_data["date_modification"] = datetime.now().replace(tzinfo=None)
        elif isinstance(processed_data["date_modification"], str):
            try:
                date_modification = datetime.fromisoformat(processed_data["date_modification"])
                # S'assurer que la date est naive (sans timezone) pour éviter les erreurs de comparaison
                if date_modification.tzinfo is not None:
                    date_modification = date_modification.replace(tzinfo=None)
                processed_data["date_modification"] = date_modification
            except ValueError:
                processed_data["date_modification"] = datetime.now().replace(tzinfo=None)

        # Gestion de date_reception si elle existe
        if "date_reception" in processed_data and isinstance(processed_data["date_reception"], str):
            try:
                date_reception = datetime.fromisoformat(processed_data["date_reception"])
                # S'assurer que la date est naive (sans timezone) pour éviter les erreurs de comparaison
                if date_reception.tzinfo is not None:
                    date_reception = date_reception.replace(tzinfo=None)
                processed_data["date_reception"] = date_reception
            except ValueError:
                processed_data["date_reception"] = None

        if "etat" not in processed_data:
            processed_data["etat"] = EtatReception.EN_COURS
        elif isinstance(processed_data["etat"], str):
            etat_found = False
            for etat in EtatReception:
                if etat.value == processed_data["etat"]:
                    processed_data["etat"] = etat
                    etat_found = True
                    break
            if not etat_found:
                processed_data["etat"] = EtatReception.EN_COURS

        # Traitement de la matière
        if "matiere" not in processed_data or processed_data["matiere"] is None:
            # Créer une matière par défaut
            processed_data["matiere"] = Matiere(code_mp="UNKNOWN", nom="Matière inconnue")
        elif isinstance(processed_data["matiere"], dict):
            try:
                processed_data["matiere"] = Matiere.from_model_dump(processed_data["matiere"])
            except Exception:
                processed_data["matiere"] = Matiere(code_mp="UNKNOWN", nom="Matière inconnue")

        # Création de l'instance
        try:
            return cls(**processed_data)
        except Exception as e:
            print(f"Erreur lors de la création de la réception: {e}")
            # En dernier recours, créer une réception avec des valeurs par défaut
            return cls(
                id=None,  # Sera généré automatiquement
                matiere=Matiere(code_mp="UNKNOWN", nom="Matière inconnue"),
                quantite=0,
                lot="",
                qualification=False,
                date_creation=datetime.now().replace(tzinfo=None),
                date_modification=datetime.now().replace(tzinfo=None),
                etat=EtatReception.EN_COURS
            )

File: interflow-frontend/src/model/reception.ts
import { Matiere, MatiereModel } from './matiere';

export enum EtatReception {
  EN_COURS = "en_cours",
  TERMINEE = "terminée",
  ANNULEE = "annulée",
  RELACHE = "relâché",
  EN_ATTENTE = "en_attente"
}

export enum TypeReception {
  EXTERNE = "externe",
  INTERNE = "interne"
}

export interface Reception {
  id?: string;
  type: TypeReception;
  matiere: Matiere;
  quantite: number;
  lot: string;
  qualification: boolean;
  date_creation: Date;
  date_modification?: Date | null;
  
  // Champs spécifiques aux réceptions internes
  ordre?: string | null;
  type_ordre?: string | null;
  statut_ordre?: string | null;
  poste?: string | null;
  statut_poste?: string | null;
  division?: string | null;
  magasin?: string | null;
  article?: string | null;
  libelle_article?: string | null;
  ref_externe?: string | null;
  description_externe?: string | null;
  quantite_ordre?: number | null;
  quantite_receptionnee?: number | null;
  udm?: string | null;
  fournisseur?: string | null;
  description_fournisseur?: string | null;
  date_reception?: Date | null;
  
  // État unifié (pour compatibilité)
  etat: EtatReception;
}

export class ReceptionModel implements Reception {
  id?: string;
  type: TypeReception;
  matiere: Matiere;
  quantite: number;
  lot: string;
  qualification: boolean;
  date_creation: Date;
  date_modification?: Date | null;
  
  // Champs spécifiques aux réceptions internes
  ordre?: string | null;
  type_ordre?: string | null;
  statut_ordre?: string | null;
  poste?: string | null;
  statut_poste?: string | null;
  division?: string | null;
  magasin?: string | null;
  article?: string | null;
  libelle_article?: string | null;
  ref_externe?: string | null;
  description_externe?: string | null;
  quantite_ordre?: number | null;
  quantite_receptionnee?: number | null;
  udm?: string | null;
  fournisseur?: string | null;
  description_fournisseur?: string | null;
  date_reception?: Date | null;
  
  // État unifié (pour compatibilité)
  etat: EtatReception;

  constructor(data: Partial<Reception>) {
    this.type = data.type || TypeReception.EXTERNE;
    this.matiere = data.matiere || new MatiereModel({ code_mp: "UNKNOWN", nom: "Matière inconnue" });
    this.quantite = data.quantite || 0;
    this.lot = data.lot || "";
    this.qualification = data.qualification || false;
    this.date_creation = data.date_creation || new Date();
    this.date_modification = data.date_modification || null;
    
    // Champs spécifiques aux réceptions internes
    this.ordre = data.ordre || null;
    this.type_ordre = data.type_ordre || null;
    this.statut_ordre = data.statut_ordre || null;
    this.poste = data.poste || null;
    this.statut_poste = data.statut_poste || null;
    this.division = data.division || null;
    this.magasin = data.magasin || null;
    this.article = data.article || null;
    this.libelle_article = data.libelle_article || null;
    this.ref_externe = data.ref_externe || null;
    this.description_externe = data.description_externe || null;
    this.quantite_ordre = data.quantite_ordre || null;
    this.quantite_receptionnee = data.quantite_receptionnee || null;
    this.udm = data.udm || null;
    this.fournisseur = data.fournisseur || null;
    this.description_fournisseur = data.description_fournisseur || null;
    this.date_reception = data.date_reception || null;
    
    this.etat = data.etat || EtatReception.EN_COURS;
    
    // Générer l'id si non fourni
    if (!data.id) {
      if (this.type === TypeReception.INTERNE && this.ordre && this.article) {
        // Utiliser ordre + article + date_reception + poste (sans heure)
        if (this.date_reception) {
          const dateStr = this.date_reception.toISOString().split('T')[0].replace(/-/g, '');
          const posteStr = this.poste || "NOPOSTE";
          this.id = `${this.ordre}_${this.article}_${dateStr}_${posteStr}`;
        } else {
          // Si pas de date de réception, utiliser la date de création
          const dateStr = this.date_creation.toISOString().split('T')[0].replace(/-/g, '');
          const posteStr = this.poste || "NOPOSTE";
          this.id = `${this.ordre}_${this.article}_${dateStr}_${posteStr}`;
        }
      } else {
        // Pour les réceptions prestataires, utiliser un id auto-généré
        this.id = crypto.randomUUID();
      }
    } else {
      this.id = data.id;
    }
    
    // Convertir les états de réceptions internes vers l'état unifié
    if (this.type === TypeReception.INTERNE && this.statut_ordre) {
      const statutUpper = this.statut_ordre.toUpperCase();
      if (statutUpper.includes("RELÂCHÉ") || statutUpper.includes("RELACHE")) {
        this.etat = EtatReception.RELACHE;
      } else if (statutUpper.includes("EN_ATTENTE") || statutUpper.includes("EN ATTENTE")) {
        this.etat = EtatReception.EN_ATTENTE;
      } else if (statutUpper.includes("TERMINÉ") || statutUpper.includes("TERMINE")) {
        this.etat = EtatReception.TERMINEE;
      } else if (statutUpper.includes("ANNULÉ") || statutUpper.includes("ANNULE")) {
        this.etat = EtatReception.ANNULEE;
      } else {
        this.etat = EtatReception.EN_COURS;
      }
    }
  }

  static fromData(data: Record<string, any>): ReceptionModel {
    const processedData = { ...data };
    
    // Valeurs par défaut pour les champs manquants
    if (!processedData.type) {
      processedData.type = TypeReception.EXTERNE;
    } else if (typeof processedData.type === 'string') {
      try {
        processedData.type = Object.values(TypeReception).find(t => t === processedData.type) || TypeReception.EXTERNE;
      } catch (error) {
        processedData.type = TypeReception.EXTERNE;
      }
    }
    
    if (!processedData.quantite) {
      processedData.quantite = 0;
    }
    
    if (!processedData.lot) {
      processedData.lot = "";
    }
    
    if (processedData.qualification === undefined) {
      processedData.qualification = false;
    }
    
    // Gestion des dates
    if (!processedData.date_creation) {
      processedData.date_creation = new Date();
    } else if (typeof processedData.date_creation === 'string') {
      try {
        processedData.date_creation = new Date(processedData.date_creation);
      } catch (error) {
        processedData.date_creation = new Date();
      }
    }
    
    if (processedData.date_modification && typeof processedData.date_modification === 'string') {
      try {
        processedData.date_modification = new Date(processedData.date_modification);
      } catch (error) {
        processedData.date_modification = null;
      }
    }
    
    if (processedData.date_reception && typeof processedData.date_reception === 'string') {
      try {
        processedData.date_reception = new Date(processedData.date_reception);
      } catch (error) {
        processedData.date_reception = null;
      }
    }
    
    // Gestion de l'état
    if (!processedData.etat) {
      processedData.etat = EtatReception.EN_COURS;
    } else if (typeof processedData.etat === 'string') {
      processedData.etat = Object.values(EtatReception).find(e => e === processedData.etat) || EtatReception.EN_COURS;
    }
    
    // Traitement de la matière
    if (!processedData.matiere) {
      processedData.matiere = new MatiereModel({ code_mp: "UNKNOWN", nom: "Matière inconnue" });
    } else if (typeof processedData.matiere === 'object') {
      try {
        processedData.matiere = MatiereModel.fromData(processedData.matiere);
      } catch (error) {
        processedData.matiere = new MatiereModel({ code_mp: "UNKNOWN", nom: "Matière inconnue" });
      }
    }
    
    try {
      return new ReceptionModel(processedData);
    } catch (error) {
      console.error("Erreur lors de la création de la réception:", error);
      // En dernier recours, créer une réception avec des valeurs par défaut
      return new ReceptionModel({
        type: TypeReception.EXTERNE,
        matiere: new MatiereModel({ code_mp: "UNKNOWN", nom: "Matière inconnue" }),
        quantite: 0,
        lot: "",
        qualification: false,
        date_creation: new Date(),
        etat: EtatReception.EN_COURS
      });
    }
  }

  toData(): Record<string, any> {
    return {
      id: this.id,
      type: this.type,
      matiere: this.matiere instanceof MatiereModel ? this.matiere.toData() : this.matiere,
      quantite: this.quantite,
      lot: this.lot,
      qualification: this.qualification,
      date_creation: this.date_creation.toISOString(),
      date_modification: this.date_modification?.toISOString() || null,
      etat: this.etat,
      // Champs spécifiques aux réceptions internes
      ordre: this.ordre,
      type_ordre: this.type_ordre,
      statut_ordre: this.statut_ordre,
      poste: this.poste,
      statut_poste: this.statut_poste,
      division: this.division,
      magasin: this.magasin,
      article: this.article,
      libelle_article: this.libelle_article,
      ref_externe: this.ref_externe,
      description_externe: this.description_externe,
      quantite_ordre: this.quantite_ordre,
      quantite_receptionnee: this.quantite_receptionnee,
      udm: this.udm,
      fournisseur: this.fournisseur,
      description_fournisseur: this.description_fournisseur,
      date_reception: this.date_reception?.toISOString() || null
    };
  }
} 